Correct CF calculations and add campaign statistical comparison

tmieslinger opened the pull request:
Hi Jule, 
based on your notebook I made some major changes to the calculation of cloud fractions that should be able to handle all variations in the cloud mask flag meanings among the datasets :D I hope...
And I added two more plots for the statistical comparison of circle mean cloud fractions for the whole campaign.

vpoertge committed:
Added the projection onto a lat/lon map to the existing specmacs notebook.

vpoertge committed:
Removed non-xarray plot and adjusted colorbar for xarray-plot

vpoertge committed:
Error in viewpath calculation. Should be delta_height/view_lh[2].

vpoertge committed:
Arrays rewritten as xarray datasets and np.einsum exchanged by xr.dot where possible.

vpoertge committed:
Using short time sequence instead of full HALO circle. Added calculation of swath width and cloud fraction plot.

tmieslinger committed:
Correct CF calculations and add campaign statistical comparison

tmieslinger commented:
Building the book won't work as AERIS is down again. You might have to rerun the checks later...

JuleRadtke commented on how_to_eurec4a/cloudmasks.md:
```diff
@@ -4,136 +4,275 @@ jupytext:
+We define some utility functions to extract (circle) cloud cover estimates from different datasets assuming that the datasets follow a certain flag meaning convention.  
+
+In particular, we define a **minimum cloud fraction** based on the cloud mask flag `most_likely_cloudy` and a **maximum cloud fraction** combining the flags `most_likely_cloudy` and `probably_cloudy`. The cloud mask datasets slightly vary in their flag definition for cloud free conditions and their handling of unvalid measurements which is taken care of by the following functions.
+
```
Great! The word `cloud fraction` still confuses me a bit when thinking about the 1D horizontal cloud masks in time, probably its better to stick to the term `cloud mask flag`, so `minimum cloud mask flag` and `maximum cloud mask flag`?

tmieslinger replied:
Hm. The `minimum` of `cloud mask flag` would be quite misleading as that minimum of possible flag values in our case would be `-1` : `unknown`. The operator `minimum` applied over `time` will provide us with a time-fraction of the cloud mask`flag`, which is exactly what we want. So I would argue that `min / max cloud fraction` is the correct term at this point. Sometimes it helps me to think of the (1D) cloud mask such as from WALES as a 2D array with the dimensions `time` and `flag`. Not sure whether this is more intuitive?

JuleRadtke replied:
Ah true, that is also confusing. I have no problem with the term cloud fraction if applied over e.g. time, but only naming it so in the functions "just" combining the cloud flags. But anyways, these are details, I think it gets clear if going along with the notebook.

JuleRadtke commented on how_to_eurec4a/cloudmasks.md:
```diff
@@ -4,136 +4,275 @@ jupytext:
+def correct_VELOX(ds):
+    return ds.assign(CF_min=ds.CF_min.where((ds.CF_min >=0 ) & (ds.CF_min <= 1)),
+                     CF_max=ds.CF_max.where((ds.CF_max >=0 ) & (ds.CF_max <= 1)))
+
```
`correct_Velox` is quite specific, but used for loading all datasets, not only Velox. Probably just add this to the function `ensure_cfminmax` without giving this an extra function with a name?

tmieslinger replied:
yes, I agree. This is a work-around as the current VELOX files have a bug. Michael will check it and hopefully correct the files so that we can delete this function again. So it's in a separate function to be easily deletable ;-)

JuleRadtke replied:
Ah that makes sense, ja!

JuleRadtke commented on how_to_eurec4a/cloudmasks.md:
```diff
@@ -4,136 +4,275 @@ jupytext:
+## Case study on February 5
+We show the cloud masks from the various instruments for an 5 minute time interval around noon on February 5.
+
+We add 2D vertical lidar and radar data, as well as 2D horizontal imager data for a better visialization of the cloud information content provided by the instruments.
```
visualization

tmieslinger replied:
typo corrected.

JuleRadtke commented on how_to_eurec4a/cloudmasks.md:
```diff
@@ -142,48 +281,138 @@ axP.yaxis.set_visible(False)
-                (1, 0.5),xycoords="axes fraction")
+## Statistical comparison
 
+we will further compare cloud mask information from all HALO flights during EUREC4A on the basis of circle flight segments. Most of the time, the HALO aircraft sampled the airmass in circles east of Barbados. We use the meta data on flight segments, extract the information on start and end time of individual circles, and derive circle-average cloud fractions.  
```
We

JuleRadtke replied:
Nice statistical plots! Here now you didn't specify for SpecMacs and Velox a smaller footprint as above, probably its worth mentioning this difference here explicitly and remember the reader of the quite different footprints.

tmieslinger replied:
I added some more information here.

JuleRadtke commented on how_to_eurec4a/cloudmasks.md:
```diff
@@ -4,136 +4,275 @@ jupytext:
+            })
+            for i in segments.values()
+        ], dim="time")
+
```
To ease the understanding which preprocessing steps have been taken on the dataset to generate which plot, I would move the functions that are only used for a specific analysis/plot to that section and leave here only the functions used generally/for all analysis in the notebook, that is moving the `midpoint` and `cf_circles` to the statistical comparison section.

tmieslinger replied:
That's a good point. I moved the functions you mentioned to the respective section.

JuleRadtke reviewed (commented):
Very nice!! I have a few small comments and a suggestion to reorder a bit the functions to ease the understanding of the various preprocessing steps to generate the specific dataset used in the different analysis/plots.

tmieslinger commented:
We still have the problem that there is a time execution limit of 2 min for the notebooks in the How-to-EUREC4A book and the current `cloud mask` notebook takes more than 20 min. 
One way would be to include it as a `.ipynb` file and with pre-compiled cells so that the plots are visable. That would have the negative effect that version control of this particular file would include the plots and would be messy.

JuleRadtke commented:
I see, and it's not so easy to change this 2 min limit? Then it seems this is the only practible solution at the moment to make it visible for the book...?

JuleRadtke committed:
update bacardi notebook with comments from the dataset contacts

tmieslinger committed:
add suggestions from review

tmieslinger committed:
add cloudmask entry to table of contents

d70-t committed:
exclude fsspec version 0.9.0 which has a bug when retrieving files via http

tmieslinger committed:
Merge pull request #45 from d70-t/fix_fsspec_http_bug

exclude fsspec version 0.9.0 which has a bug when retrieving files via http

tmieslinger committed:
Merge remote-tracking branch 'upstream/master' into cloudmask

tmieslinger committed:
Correct non-consecutive header level increase

d70-t committed:
Merge remote-tracking branch 'upstream/master' into specmacs-projection

d70-t committed:
specmacs rewording

This commit contains a bit of rewording for the specmacs cloudmask
notebook (both, in terms of text and more xarray usage).

d70-t committed:
removed breaking type annotation

d70-t committed:
Apply suggestions from code review

Co-authored-by: vpoertge <41779856+vpoertge@users.noreply.github.com>

d70-t committed:
rewording of altitude and height

d70-t committed:
added introductory words about cloudmask from SWIR

d70-t committed:
added "cth" to specmacs_lat_lon drawing

d70-t committed:
typos

vpoertge committed:
Merge pull request #1 from d70-t/specmacs_rewording

Some small rewording to the projection notebook

d70-t committed:
fixed accidental stroke in specmacs_lat_lon drawing

d70-t committed:
adjusted WALES assertations in HAMP comparison

d70-t committed:
refactore WALES notebook to correctly account for missing data

d70-t committed:
Apply suggestions from code review

Co-authored-by: Theresa Mieslinger <41184617+tmieslinger@users.noreply.github.com>

d70-t committed:
increased notebook timeout

The interactive flight track notebook is always really close to the limit, this should give a little more room.

tmieslinger committed:
Merge pull request #46 from d70-t/wales_update

Wales update

d70-t committed:
Merge remote-tracking branch 'upstream/master' into specmacs-projection

d70-t committed:
Merge pull request #41 from vpoertge/specmacs-projection

Specmacs cloud fraction and projection onto map

d70-t commented on how_to_eurec4a/cloudmasks.md:
````diff
@@ -4,136 +4,261 @@ jupytext:
 
 ```{code-cell} ipython3
-s = slice(datetime.datetime(2020, 2, 5, 11, 22, 0), datetime.datetime(2020, 2, 5, 11, 27, 0))
+def most_frequent(da):
+    flag, n = np.unique(da.values, return_counts=True)
+    return xr.DataArray(int(flag[np.argmax(n)]), dims=())
+
+def select_specmacs_cloudmask(ds):
+    specmacs = ds.sel(angle=slice(.3, -.3))
+    return ds.assign({"cloud_mask": specmacs.cloud_mask.groupby("time").map(most_frequent),
+                      "CF_min": cfmin(specmacs),
+                      "CF_max": cfmax(specmacs)})
+
+def select_velox_cloudmask(ds):
+    xmid = ds.x.size // 2
+    ymid = ds.y.size // 2
+    velox = ds.isel(x=slice(xmid - 5, xmid + 5), y=slice(ymid - 5, ymid + 5))
+    return ds.assign({"cloud_mask": velox.cloud_mask.groupby("time").map(most_frequent),
+                      "CF_min": cfmin(velox),
+                      "CF_max": cfmax(velox)})
````
This one should provide quite a bit of speed-up.

Take-home-message: `groupby` is really slow...
```suggestion
def most_frequent_flag(var, dims):
    flags = xr.DataArray(var.flag_values, dims=("__internal__flags__"))
    flag_indices = (var == flags).sum(dims).argmax("__internal__flags__")
    return xr.DataArray(var.flag_values[flag_indices.data],
                        dims=flag_indices.dims,
                        attrs=var.attrs)

def select_specmacs_cloudmask(ds):
    specmacs = ds.sel(angle=slice(.3, -.3))
    return ds.assign({"cloud_mask": most_frequent_flag(specmacs.cloud_mask, "angle"),
                      "CF_min": cfmin(specmacs),
                      "CF_max": cfmax(specmacs)})

def select_velox_cloudmask(ds):
    xmid = ds.x.size // 2
    ymid = ds.y.size // 2
    velox = ds.isel(x=slice(xmid - 5, xmid + 5), y=slice(ymid - 5, ymid + 5))
    return ds.assign({"cloud_mask": most_frequent_flag(velox.cloud_mask, ("x", "y")),
                      "CF_min": cfmin(velox),
                      "CF_max": cfmax(velox)})
```

d70-t commented on how_to_eurec4a/cloudmasks.md:
```diff
@@ -4,136 +4,261 @@ jupytext:
+        ds = ds.assign(CF_max=cfmax)
+    return correct_VELOX(ds)
+
+def load_cloudmask_dataset(cat_item):
+    return ensure_cfminmax(xr.concat([v.get().to_dask().chunk()
+                                      for _, v in cat_item.items()], dim="time"))
```
This one loads the individual datasets (and in particular coordinate indices) for each flight day in parallel. That should have a bit of an effect on data loading speed.
```suggestion
from multiprocessing.pool import ThreadPool

def load_cloudmask_dataset(cat_item):
    # load in parallel as this function is mainly limited by the network roundtrip time
    p = ThreadPool(20)
    return ensure_cfminmax(xr.concat(list(p.map(lambda v: v.get().to_dask().chunk(), cat_item.values())),
                                     dim="time"))
```

d70-t reviewed (commented):
Just two suggestions for notebook speedup.

RobertPincus committed:
Add multi-platform datasets starting with JOANNE dropsondes

JuleRadtke committed:
Merge branch 'master' into bacardi_update

d70-t committed:
Merge pull request #44 from JuleRadtke/bacardi_update

update bacardi notebook with comments from the dataset contacts

tmieslinger committed:
Merge pull request #48 from RobertPincus/add-multi-platform-products

Add multi-platform datasets starting with JOANNE dropsondes

tmieslinger committed:
Update how_to_eurec4a/cloudmasks.md

Co-authored-by: Tobias Kölling <tobias.koelling@mpimet.mpg.de>

tmieslinger committed:
speed up how_to_eurec4a/cloudmasks.md

Co-authored-by: Tobias Kölling <tobias.koelling@mpimet.mpg.de>

tmieslinger committed:
Merge remote-tracking branch 'upstream/master' into cloudmask

tmieslinger committed:
Add `CF_min` to statistic plots
